docs/architecture/grpc-for-wcf-developers/application-performance-management.md

```diff
@@ -0,0 +1,227 @@
+| Meter       | Measures the rate at which an event occurs in various time spans. |
+| Timer       | Tracks the duration of events and the rate at which it occurs, stored as a histogram. |
+
+Using *App Metrics*, an `IMetrics` interface can be obtained via dependency injection and used to record any of these metrics for a gRPC service. The following example shows how to count the number of `Get` requests made over time:
+
+```csharp
+public class StockData : Stocks.StocksBase
+{
+    private static readonly CounterOptions GetRequestCounter = new CounterOptions
+    {
+        Name = "StockData_Get_Requests",
+        MeasurementUnit = Unit.Calls
+    };
+
+    private readonly IStockRepository _repository;
+    private readonly IMetrics _metrics;
+
+    public StockData(IStockRepository repository, IMetrics metrics)
+    {
+        _repository = repository;
+        _metrics = metrics;
+    }
+
+    public override async Task<GetResponse> Get(GetRequest request, ServerCallContext context)
+    {
+        _metrics.Measure.Counter.Increment(GetRequestCounter);
+
+        // Serve request...
+    }
+}
+```
+
+### Storing and visualizing metrics data
+
+The best way to store metrics data is in a *time-series database*, a specialized data store designed to record numerical data series marked with timestamps. The most popular of these databases are [Prometheus](https://prometheus.io/) and [InfluxDB](https://www.influxdata.com/products/influxdb-overview/). Microsoft Azure also provides dedicated metrics storage through the [Azure Monitor](https://docs.microsoft.com/azure/azure-monitor/overview) service.
+
+The current go-to solution for visualizing metrics data is [Grafana](https://grafana.com), which works with a wide range of storage providers including Azure Monitor, InfluxDB and Prometheus. The following image shows an example Grafana dashboard that displays metrics from the Linkerd service mesh running the StockData sample:
+
+![Grafana dashboard](media/application-performance-management/grafana-screenshot.png)
+
+### Metrics-based alerting
+
+The numerical nature of metrics data means that it's ideally suited to drive alerting systems, notifying developers or support engineers when a value falls outside of some defined tolerance. The platforms already mentioned all provide support for alerting via a range of options, including emails, text messages, or in-dashboard visualizations.
+
+## Distributed tracing
+
+*Distributed tracing* is a relatively recent development in monitoring, which has arisen from the increasing use of microservices and distributed architectures. A single request from a client browser, application, or device may be broken down into many steps and subrequests and involve the use of many services across a network. This makes it difficult to correlate log messages and metrics with the specific request that triggered them.
```

Comment:
WCF has supported end to end tracing since at least .NET 4.0. We can trace from a client to a service, to other service and visualize the entire call tree. It's also been possible to add your own tracing. The thing that the new tracing systems have is that everyone is getting on board now so things like HttpClient are participating and there's systems to centrally log your tracing. But if you grabbed the log files off each system, you could use ServiceTraceViewer to see your entire end-to-end call distributed across all your WCF components. Here's a [blog post](https://social.technet.microsoft.com/wiki/contents/articles/8068.wcf-end-to-end-tracing-simplified.aspx) showing how it works.

Reply:
Added a link to https://docs.microsoft.com/en-us/dotnet/framework/wcf/diagnostics/tracing/end-to-end-tracing with a note that it's similar. Thanks.